include/jwt-cpp/jwt.h

```diff
@@ -2699,130 +2707,133 @@ namespace jwt {
-				error_msg += name;
-				error_msg += "\" not found";
-				throw std::runtime_error(error_msg.c_str());
+		basic_claim_t get_jwk_claim(const typename json_traits::string_type& name) const {
+			if (!has_jwk_claim(name)) {
+				throw std::runtime_error("claim not found");
 			}
-			return jwks_claims.at(name);
+			return jwk_claims.at(name);
 		}
 
-		bool empty() const noexcept { return jwks_claims.empty(); }
-
-	protected:
-		JWT_CLAIM_EXPLICIT jwks(const typename json_traits::value_type& claim) {
-			for (const auto& e : json_traits::as_object(claim)) {
-				jwks_claims.emplace(e.first, basic_claim_t(e.second));
-			}
-		}
+		bool empty() const noexcept { return jwk_claims.empty(); }
 
 	private:
-		std::unordered_map<typename json_traits::string_type, basic_claim_t> jwks_claims;
+		std::unordered_map<typename json_traits::string_type, basic_claim_t> jwk_claims;
 	};
 
 	template<typename json_traits>
-	class jwks_keys {
+	class jwks {
 		using basic_claim_t = basic_claim<json_traits>;
-		using jwks_t = jwks<json_traits>;
+		using jwk_t = jwk<json_traits>;
 
 	public:
 
-		JWT_CLAIM_EXPLICIT jwks_keys(const typename json_traits::string_type& jwks_keys_token) {
+		JWT_CLAIM_EXPLICIT jwks(const typename json_traits::string_type& jwks_token) {
 			auto parse_claims = [](const typename json_traits::string_type& str) {
-				std::unordered_map<typename json_traits::string_type, jwks_t> res;
+				std::unordered_map<typename json_traits::string_type, jwk_t> res;
 				typename json_traits::value_type val;
 				if (!json_traits::parse(val, str))
 					throw std::runtime_error("Invalid json");
 
 				for (const auto& k : json_traits::as_object(val)) {
+					size_t id = 0;
 					for (const typename json_traits::value_type& item : json_traits::as_array(k.second)) {
-						jwks_t jwks_entry(item);
-						res.emplace(jwks_entry.get_key_id(), std::move(jwks_entry));
+						jwk_t jwk_entry(item);
+						if (jwk_entry.has_key_id())
+							res.emplace(jwk_entry.get_key_id(), std::move(jwk_entry));
+						else
+							res.emplace(std::to_string(id), std::move(jwk_entry));
+						++id;
 					}
 				}
 
 				return res;
 			};
 
-			jwks_keys_claims = parse_claims(jwks_keys_token);
+			jwks_claims = parse_claims(jwks_token);
 		}
 
 		/**
-		 * Get jwks
-		 * \return Requested jwks by key_id
-		 * \throw std::runtime_error If jwks was not present
+		 * Get jwk
+		 * \return Requested jwk by key_id
+		 * \throw std::runtime_error If jwk was not present
 		 */
-		jwks_t get_jwks(const typename json_traits::string_type& key_id) const {
-			if (!has_jwks(key_id)) {
-				typename json_traits::string_type error_msg = "jwks with key id \"";
+		jwk_t get_jwk(const typename json_traits::string_type& key_id) const {
+			if (!has_jwk(key_id)) {
+				typename json_traits::string_type error_msg = "jwk with key id \"";
 				error_msg += key_id;
 				error_msg += "\" not found";
 				throw std::runtime_error(error_msg.c_str());
```

Comment:
the caller knows the kid they passed, no need to repeat it

```suggestion
				throw std::runtime_error("jwk with matching key id not found");
```